MFTF Automation for: User locates licensed and uploaded image inside Media Gallery

drpayyne opened the pull request:
### Description
-  Added MFTF test for the below mentioned issue

### Fixed Issues
-  magento/adobe-stock-integration#444: [MFTF] User locates licensed and uploaded image inside Media Gallery

### Manual testing scenarios

-  Add `magento/ADOBE_STOCK_LICENSED_IMAGE` credential with the image ID of a licensed image
-  `$ vendor/bin/mftf run:test AdminAdobeStockSavedLicensedImageLocateTest --remove`

drpayyne committed:
Resolved issue #444

drpayyne committed:
Resolved issue #444

diazwatson commented:
Hi @drpayyne, thanks for your contribution.
While running your test I get the following error:

```
Time: 2.99 minutes, Memory: 12.00MB

There was 1 error:

---------
1) AdminAdobeStockSavedLicensedImageLocateTestCest: Admin adobe stock saved licensed image locate test
 Test  tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:AdminAdobeStockSavedLicensedImageLocateTest

  [Facebook\WebDriver\Exception\UnrecognizedExceptionException] element click intercepted: Element <button id="search_adobe_stock" title="Search Adobe Stock" type="button" class="action-default scalable action-secondary" onclick="jQuery(&quot;#adobe-stock-images-search-modal&quot;).trigger(&quot;openModal&quot;);" data-ui-id="wysiwyg-images-content-search-adobe-stock-button">...</button> is not clickable at point (532, 126). Other element would receive the click: <div class="action-buttons preview-buttons">...</div>
  (Session info: chrome=78.0.3904.87)
  (Driver info: chromedriver=78.0.3904.70 (edb9c9f3de0247fd912a77b7f6cae7447f6d3ad5-refs/branch-heads/3904@{#800}),platform=Mac OS X 10.15.1 x86_64)


Scenario Steps:

 59. $I->click("[data-ui-id=wysiwyg-images-content-search-adobe-stock-button]") at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:82
 58. // Entering Action Group [openAdobeStockPanel] AdminAdobeStockOpenPanelFromMediaGalleryActionGroup
 57. $I->saveScreenshot() at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:113
 56. $I->grabAttributeFrom("//button[@class='action-secondary']//spa...","name") at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:136
 55. // Exiting Action Group [expandLicensedImage] AdminAdobeStockExpandImagePreviewActionGroup
 54. $I->waitForLoadingMaskToDisappear() at tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:134

#1  /Users/rdiaz/Sites/adobestock/vendor/facebook/webdriver/lib/Exception/WebDriverException.php:158
#2  /Users/rdiaz/Sites/adobestock/vendor/facebook/webdriver/lib/Remote/HttpCommandExecutor.php:331
#3  /Users/rdiaz/Sites/adobestock/vendor/facebook/webdriver/lib/Remote/RemoteWebDriver.php:565
#4  /Users/rdiaz/Sites/adobestock/vendor/facebook/webdriver/lib/Remote/RemoteExecuteMethod.php:40
#5  /Users/rdiaz/Sites/adobestock/vendor/facebook/webdriver/lib/Remote/RemoteWebElement.php:81
#6  Codeception\Module\WebDriver->click
#7  /Users/rdiaz/Sites/adobestock/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/_generated/AcceptanceTesterActions.php:1136
#8  /Users/rdiaz/Sites/adobestock/dev/tests/acceptance/tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:82
#9  Magento\AcceptanceTest\_default\Backend\AdminAdobeStockSavedLicensedImageLocateTestCest->_after

ERRORS!
Tests: 1, Assertions: 1, Errors: 1.
```

Have you seen this before?
Note: in order to run this test in isolation I had to comment out the following lines:
```
<group value="adobe_stock_integration_license"/>
<group value="adobe_stock_integration"/>
```

drpayyne commented:
@diazwatson No I haven't got that error. Let me try in my local once again.

drpayyne commented:
@diazwatson I think it's due to the _Locate_ button image fix not being present here. (https://github.com/magento/adobe-stock-integration/pull/637#issuecomment-549814965) Testing with the fix in my local now.

drpayyne committed:
Issue-444 bug fix and code refactor

drpayyne committed:
Issue-444 bug fix and code refactor

drpayyne commented:
Hi @diazwatson. I've fixed it now, thanks. Also note that I've changed environment variable name for more clarity due to multiple similar MFTF tests present for Adobe Stock image licensing.

drpayyne committed:
Made after node stateless

diazwatson commented:
Great! thanks for your effort @drpayyne I am going to re test it now

diazwatson commented:
Just ran the test and it complete successfully. 
```
AdminAdobeStockSavedLicensedImageLocateTestCest: Admin adobe stock saved licensed image locate test
Signature: Magento\AcceptanceTest\_default\Backend\AdminAdobeStockSavedLicensedImageLocateTestCest:AdminAdobeStockSavedLicensedImageLocateTest
Test: tests/functional/Magento/FunctionalTest/_generated/default/AdminAdobeStockSavedLicensedImageLocateTestCest.php:AdminAdobeStockSavedLicensedImageLocateTest
Scenario --
[loginAsAdmin] LoginAsAdmin
  [navigateToAdmin] am on page "/admin/admin"
  [fillUsername] fill field "#username","admin"
  [fillPassword] fill field "#login","123123q"
  [clickLogin] click ".actions .action-primary"
  [clickLoginWaitForPageLoad] wait for page load 30
  [closeAdminNotification] close admin notification
[openMediaGalleryForPage] AdminOpenMediaGalleryForPageNoEditorActionGroup
  [amOnPagePagesGrid] am on page "/admin/cms/page"
  [waitForPageLoad] wait for page load 10
  [clickAddNewPage] click "#add"
  [clickAddNewPageWaitForPageLoad] wait for page load 30
  [clickExpandContent] click "div[data-index=content]"
  [waitForInsertImageButton] wait for element visible ".scalable.action-add-image.plugin",10
  [clickInsertImage] click ".scalable.action-add-image.plugin"
  [] wait for initial media gallery load, where the gallery chrome loads (and triggers loading modal)
  [waitForMediaGalleryInitialLoad] wait for page load 10
  [] javascript ERROR(SEVERE) - https://adobestock.test/static/version1572277603/adminhtml/Magento/backend/en_US/mage/adminhtml/browser.js 91:47 Uncaught T
  [] wait for second media gallery load, where the gallery images load (and triggers loading modal once more)
  [waitForMediaGallerySecondaryLoad] wait for page load 10
[openAdobeStockPanel] AdminAdobeStockOpenPanelFromMediaGalleryActionGroup
  [openAdobeStockPanel] click "[data-ui-id=wysiwyg-images-content-search-adobe-stock-button]"
  [waitForAdobeStockPanelOpen] wait for page load 10
[clickOnSignIn] AdminAdobeStockClickSignInActionGroup
  [clickOnSignInButton] click ".adobe-sign-in-button"
  [waitForLoad] wait for page load 10
[fillUserCredentials] AdminAdobeStockImsPopupSignInFillUserDataActionGroup
  [switchToWindow] switch to window
  [fillUserEmail] fill field "#adobeid_username","diazwatsonre@gmail.com"
  [clickPasswordField] click "#adobeid_password"
  [waitForUserEmailToLoad] wait for page load 10
  [fillUserPassword] fill field "#adobeid_password","47b4G3rQEuZ0"
[clickSignInImsPopup] AdminAdobeStockImsPopupClickSignInActionGroup
  [clickOnSignInButton] click "#sign_in"
  [switchToTab] switch to next tab
  [waitForLoad] wait for page load 10
[assertUserLoggedIn] AdminAdobeStockAssertUserLoggedActionGroup
  [seeUserIcon] see element ".adobe-profile-image-small"
[searchForLicensedImage] AdminSearchImagesOnModalActionGroup
  [fillQueryInput] fill field "input.data-grid-search-control","176476298"
  [applySearch] click "#words + button.action-submit"
  [waitForSearchForQuery] wait for loading mask to disappear
[expandLicensedImage] AdminAdobeStockExpandImagePreviewActionGroup
  [clickOnThumbnail] click "aside.adobe-stock-modal [data-role=grid-wrapper] img[data-role=thumbnail]"
  [waitForImagePreviewToExpand] wait for loading mask to disappear
[grabImageFileName] grab attribute from "//button[@class='action-secondary']//span[text()='Locate']","name"
[clickLocate] click "//button[@class='action-secondary']//span[text()='Locate']"
[waitForMediaGalleryOpen] wait for page load 10
[assertSavedImage] see element "//img[contains(@alt,'motorcycle-wheels-there-is-a-wom.jpeg')]"
[openMediaGalleryForPage] AdminOpenMediaGalleryForPageNoEditorActionGroup
  [amOnPagePagesGrid] am on page "/admin/cms/page"
  [waitForPageLoad] wait for page load 10
  [clickAddNewPage] click "#add"
  [clickAddNewPageWaitForPageLoad] wait for page load 30
  [clickExpandContent] click "div[data-index=content]"
  [waitForInsertImageButton] wait for element visible ".scalable.action-add-image.plugin",10
  [clickInsertImage] click ".scalable.action-add-image.plugin"
  [] wait for initial media gallery load, where the gallery chrome loads (and triggers loading modal)
  [waitForMediaGalleryInitialLoad] wait for page load 10
  [] javascript ERROR(SEVERE) - https://adobestock.test/static/version1572277603/adminhtml/Magento/backend/en_US/mage/adminhtml/browser.js 91:47 Uncaught T
  [] wait for second media gallery load, where the gallery images load (and triggers loading modal once more)
  [waitForMediaGallerySecondaryLoad] wait for page load 10
[openAdobeStockPanel] AdminAdobeStockOpenPanelFromMediaGalleryActionGroup
  [openAdobeStockPanel] click "[data-ui-id=wysiwyg-images-content-search-adobe-stock-button]"
  [waitForAdobeStockPanelOpen] wait for page load 10
[adobeLogout] AdminAdobeStockUserSignOutActionGroup
  [clickOnUserNameButton] click ".adobe-user-name"
  [clickOnUserSignOutButton] click ".adobe-sign-out-button"
  [waitForLoad] wait for page load 10
[assertAdobeUserLoggedOut] AdminAdobeStockAssertUserNotLoggedActionGroup
  [userIconIsNotVisible] don't see element ".adobe-profile-image-small"
  [signInLinkIsVisible] see element ".adobe-sign-in-button"
[resetAdminDataGridToDefaultView] resetAdminDataGridToDefaultView
  [openViewBookmarks] click "div.admin__data-grid-action-bookmarks button[data-bind='toggleCollapsible']"
  [openViewBookmarksWaitForPageLoad] wait for page load 30
  [selectDefaultGridView] click "//div[contains(@class, 'admin__data-grid-action-bookmarks')]/ul/li/div/a[text() = 'Default View']"
  [selectDefaultGridViewWaitForPageLoad] wait for page load 30
  [seeDefaultViewSelected] see "Default View","div.admin__data-grid-action-bookmarks button[data-bind='toggleCollapsible']"
  [seeDefaultViewSelectedWaitForPageLoad] wait for page load 30
[adminLogout] logout
  [amOnLogoutPage] am on page "/admin/admin/auth/logout/"
 PASSED

--------------------------------------------------------------------------------
200x DEPRECATION: Calling the "Symfony\Component\EventDispatcher\EventDispatcherInterface::dispatch()" method with the event name as the first argument is deprecated since Symfony 4.3, pass it as the second argument and provide the event object as the first argument instead. /Users/rdiaz/Sites/adobestock/vendor/symfony/event-dispatcher/EventDispatcher.php:58


Time: 7.32 minutes, Memory: 8.00MB

OK (1 test, 5 assertions)
```
While during the test I noticed few repetitive and maybe unnecessary (for the purpose of this test) steps, I believe this is a side effect of the action groups.

Other than that, I think is all fine.

@filmaj  @sivaschenko can you spot any way this can be improved?

filmaj commented:
Damn @diazwatson , you also have long test run times, and you have the same "JS error severe" in your MFTF output as others. You should check in on #652 as that JS error may be the root of slow test runs on certain systems.

drpayyne commented:
@diazwatson thank you, but which parts in the action group are unnecessary? Just for the sake of knowledge. Maybe I can improve it if possible.

filmaj reviewed (changes requested):
So as we've discussed in different issues / PRs, I think this test needs modification to expand the `before` and `after` clauses:

- before the test assertions execute, we'll need to save a licensed image to the media gallery
- after the test, we'll need to delete the saved + licensed image from the media gallery

drpayyne committed:
Made MFTF test stateless

drpayyne commented:
Hey @filmaj, I have added the necessary modifications. Thanks!

filmaj commented:
Hmm.. for me, in the `before`, it fails at this point:

```
[searchForLicensedImage] AdminSearchImagesOnModalActionGroup
  [fillQueryInput] fill field "input.data-grid-search-control","286863296"
  [applySearch] click "#words + button.action-submit"
  [waitForSearchForQuery] wait for loading mask to disappear
[expandLicensedImage] AdminAdobeStockExpandImagePreviewActionGroup
  [clickOnThumbnail] click "aside.adobe-stock-modal [data-role=grid-wrapper] img[data-role=thumbnail]"
  [waitForImagePreviewToExpand] wait for loading mask to disappear
[clickSave] click "//button[@class='action-default primary']//span[text()='Save']"
[saveScreenshot] save screenshot
```

.. then explodes in the `after` since the  `before` did not finish completing. I will try to dig into this a bit deeper.

drpayyne commented:
@filmaj, a probable test fail at that particular button click might be that the image isn't licensed, so the button text was _License and Save_, or maybe the image was already saved, so only _Locate_ button was present.

filmaj commented:
@drpayyne I was looking at the test  as it  was running, can confirm  the button text was just  'Save'... I'm going to need to comment out the `after` clause and run in again to try to extract more failure details from the `before` clause, as the `after` clause's `deleteImage` action blows up and covers over any earlier failures.

Will report back when I have time to look into this.. not sure if I'll get to it this weekend, may have  to wait until Monday.

drpayyne commented:
@filmaj, no issues. I shall try on my local too asap.

drpayyne commented:
Hi @filmaj, I ran the test and yes I can confirm the button text is _Save_ but this is the actual error thrown: ` element click intercepted: Element <span>...</span> is not clickable at point (1147, 137). Other element would receive the click: <div class="preview-row-content">...</div>`.  
`<div class="preview-row-content">` is the container for the image preview view. Not sure why this is being clicked, will debug further.

filmaj commented:
Can also confirm that local media gallery was empty in my runs here, that is, no images were saved before the test ran.

drpayyne committed:
Fixed incorrect selector

drpayyne commented on AdobeStockImageAdminUi/Test/Mftf/Section/AdobeStockImagePreviewSection.xml:
```diff
@@ -8,7 +8,9 @@
-        <element name="save" type="block" selector="//button[@class='action-secondary']/span[contains(.,'Save Preview')]"/>
+        <element name="save" type="button" selector="//button[@class='action-secondary']//span[text()='Save Preview']"/>
+        <element name="locateImage" type="button" selector="//button[@class='action-secondary']//span[text()='Locate']"/>
+        <element name="saveLicensedImage" type="button" selector="//div[@class='actions']/descendant::span[text()='Save']"/>
```
Hey @filmaj, sorry for the delay but I finally resolved the issue we encountered in this. 

-  Turns out that for some reason selenium isn't able to select the _Save_ button and clicks a totally different location (I'm thinking it might be due the fact that this button has multiple classes and is similar to _License and Save_, but I might be wrong). I fixed it by giving the _Save_ button a more specific selector and **the test is successful now**.

filmaj replied:
Thanks! This part now passes for me locally.

filmaj commented on AdobeStockImageAdminUi/Test/Mftf/Test/AdminAdobeStockSavedLicensedImageLocateTest.xml:
```diff
@@ -0,0 +1,61 @@
+            <argument name="query" value="{{AdobeStockConfigDataLicensedImage.value}}"/>
+        </actionGroup>
+        <actionGroup ref="AdminAdobeStockExpandImagePreviewActionGroup" stepKey="expandLicensedImage"/>
+        <grabAttributeFrom selector="{{AdobeStockImagePreviewSection.locateImage}}" userInput="name" stepKey="grabImageFileName"/>
```
This line returns the empty string in my local test r un, as the span element containing the Locate text does not have a `name` attribute. This results in the rest of the test being unable to compare against the file name. Here's a snippet of this happening in my test run locally (note the empty string used with the `alt` attribute):

```
[searchForLicensedImage] AdminSearchImagesOnModalActionGroup
  [fillQueryInput] fill secret field "input.data-grid-search-control","84Gol3inT8K17CaG7uHIhg=="
  [applySearch] click "#words + button.action-submit"
  [waitForSearchForQuery] wait for loading mask to disappear
[expandLicensedImage] AdminAdobeStockExpandImagePreviewActionGroup
  [clickOnThumbnail] click "aside.adobe-stock-modal [data-role=grid-wrapper] img[data-role=thumbnail]"
  [waitForImagePreviewToExpand] wait for loading mask to disappear
[grabImageFileName] grab attribute from "//button[@class='action-secondary']//span[text()='Locate']","name"
[clickLocate] click "//button[@class='action-secondary']//span[text()='Locate']"
[waitForMediaGalleryOpen] wait for page load 10
[assertSavedImage] see element "//img[contains(@alt,'')]"
[deleteImageActionGroup] AdminMediaGalleryDeleteImage
  [seeImageToDelete] see element "//img[contains(@alt,'')]"
  [clickImageToDelete] click "//img[contains(@alt,'')]"
[saveScreenshot] save screenshot
```

drpayyne replied:
Hey @filmaj, I just tested it locally and can confirm that this test just passed and the `name` attribute with the filename is present in the _Locate_ button. :thinking: Could you check if `actions.js` file in this PR is loaded in your browser with the method `getImageName()`?

filmaj replied:
As you pointed  out to me on Slack, I must  not be in developer  mode and client side assets a re being cached in my system. After doing  that song and dance, I get a different error around finding  the image in the media gallery once clicking the Locate button:

```
1) AdminAdobeStockSavedLicensedImageLocateTestCest: Admin adobe stock saved licensed image locate test
 Test  tests/functional/Magento/FunctionalTest/_generated/adobe_stock_integration_suite/AdminAdobeStockSavedLicensedImageLocateTestCest.php:AdminAdobeStockSavedLicensedImageLocateTest
 Step  See element "//img[contains(@alt,'hygge-concept-with-cat-book-and.jpeg')]"
 Fail  Failed asserting that an array is not empty.
```

I think it may be an issue with my environment. When clicking the Locate button, it takes a long time for the media gallery to load up on my local instance...

filmaj reviewed (changes requested):
The `before` bits  now save the image locally, which is great, but extracting the saved file name using the Locate span's `name` attribute fails for me - see my comments below. So this still needs a little more work.

sivaschenko committed:
Merge branch 'develop' of github.com:magento/adobe-stock-integration into issue-444

sivaschenko committed:
magento/adobe-stock-integration#639: Fixed locate selector, reverted js-html changes, refactored the test

ghost commented:
Hi @drpayyne, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+adobe-stock-integration+639), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
